kernel/main.c

```diff
@@ -340,6 +353,116 @@ async_init(void)
+    * map them.
+    */
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+      get_mem_region(i, &ma);
+
+      if (ma.type != TILCK_BOOT_EFI_RUNTIME_CODE && ma.type != TILCK_BOOT_EFI_RUNTIME_DATA)
+         continue;
+
+      const ulong pbegin = (ulong)ma.addr;
+      const ulong pend = MIN((ulong)(ma.addr+ma.len), (ulong)LINEAR_MAPPING_SIZE);
+
+      const size_t page_count = (pend - pbegin) >> PAGE_SHIFT;
+
+      ulong *vbegin = TO_PTR(pbegin);
+
+      /*
+       * SetupVirtualAddressMap() requires the runtime data
+       * as well as runtime code to be writable. So use
+       * `PAGING_FL_RW` for both.
+       */
+      size_t count =
+         map_pages(get_kernel_pdir(),
+                   vbegin,
+                   pbegin,
+                   page_count,
+                   PAGING_FL_BIG_PAGES_ALLOWED |
+                   PAGING_FL_RW);
+
+      if (count != page_count)
+         panic("kmalloc: unable to map regions in the virtual space");
+
+      num_entries++;
+   }
+
+   EFI_MEMORY_DESCRIPTOR *virt_map = kmalloc((size_t)num_entries);
+   num_entries = 0;
+
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+      get_mem_region(i, &ma);
+
+      if (ma.type != TILCK_BOOT_EFI_RUNTIME_CODE && ma.type != TILCK_BOOT_EFI_RUNTIME_DATA)
+         continue;
+
+      EFI_MEMORY_DESCRIPTOR desc;
+
+      desc.Type = (ma.type == TILCK_BOOT_EFI_RUNTIME_CODE) ? EfiRuntimeServicesCode : EfiRuntimeServicesData;
+      desc.PhysicalStart = (ulong)ma.addr;
+      desc.VirtualStart = (ulong)KERNEL_PA_TO_VA(ma.addr);
+      desc.NumberOfPages = ma.len / PAGE_SIZE;
+      desc.Attribute = EFI_MEMORY_RUNTIME;
+
+      virt_map[num_entries++] = desc;
+
+      printk("Type: %d\tPhysical Address: %lld\t Virtual Address: %lld\n", desc.Type, desc.PhysicalStart, desc.VirtualStart);
+   }
+
+   printk("Virtual map: %p\n", virt_map);
+
+   printk("Time %p\n", RT->GetTime);
+   printk("SVAM %p\n", RT->SetVirtualAddressMap);
+
+   UINTN mem_map_size = num_entries * sizeof(EFI_MEMORY_DESCRIPTOR);
+   UINTN desc_size = sizeof(EFI_MEMORY_DESCRIPTOR);
+   UINT32 desc_version = EFI_MEMORY_DESCRIPTOR_VERSION;
+
+   EFI_STATUS status = RT->SetVirtualAddressMap(mem_map_size, desc_size, desc_version, virt_map);
+
+   printk("%d\n", status);
+
+   EFI_TIME Time;
+   RT->GetTime(&Time, NULL);
+
+   printk("Year %d, Month %d, Day %d\n", Time.Year, Time.Month, Time.Day);
+
+   for (int i = 0; i < get_mem_regions_count(); i++) {
+      get_mem_region(i, &ma);
+
+      if (ma.type != TILCK_BOOT_EFI_RUNTIME_CODE && ma.type != TILCK_BOOT_EFI_RUNTIME_DATA)
+         continue;
+
+      const ulong pbegin = (ulong)ma.addr;
+      const ulong pend = MIN((ulong)(ma.addr+ma.len), (ulong)LINEAR_MAPPING_SIZE);
+
+      const size_t page_count = (pend - pbegin) >> PAGE_SHIFT;
+
+      ulong *vbegin = TO_PTR(pbegin);
+
+      unmap_pages(get_kernel_pdir(),
+                  vbegin,
+                  page_count,
+                  false);
+   }
+}
```

Comment:
Here, at the very end, you should try (but not part of the change, just for tests) to call GetTime() and see if it works properly.